gpflow/monitor.py

```diff
@@ -0,0 +1,245 @@
+
+
+from abc import ABC, abstractmethod
+from io import BytesIO
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import numpy as np
+import tensorflow as tf
+from matplotlib.backends.backend_agg import FigureCanvasAgg
+from matplotlib.figure import Axes, Figure
+
+from .models import BayesianModel
+from .utilities import parameter_dict
+
+
+class MonitorTask(ABC):
+    """
+    A base class for a monitoring task.
+
+    All monitoring tasks are callable objects.
+    A descendant class must implement the `run` method, which is the body of the monitoring task.
+    """
+
+    def __init__(self, period: int = 1):
+        """
+        :param period: defines how often to run the task; it will execute every `period`th step.
+            For large values of `period` the task will be less frequently run.
+        """
+        self.current_step = 0
+        self.period = period
+
+    def __call__(self, step: int, **kwargs):
+        """
+        It calls the 'run' function and sets the current step.
+
+        :param step: current step in the optimisation.
+        :param kwargs: additional key-word arguments that can be passed
+            to the `run` method of the task. This is in particular handy for
+            passing keyword argument to the callback of `ScalarToTensorBoard`.
+        """
+        self.current_step = tf.cast(step, tf.int64)
+        if step % self.period == 0:
+            self.run(**kwargs)
+
+    @abstractmethod
+    def run(self, **kwargs):
+        """
+        Implements the task to be executed on __call__.
+        The current step is available through `self.current_step`.
+
+        :param kwargs: keyword arguments available to the run method.
+        """
+        raise NotImplementedError
+
+
+class ToTensorBoard(MonitorTask):
+    def __init__(self, log_dir: str, period: int = 1):
+        """
+        :param log_dir: directory in which to store the tensorboard files.
+            Can be a nested. E.g. ./logs/my_run/
+        :param period: defines how often to run the task; it will execute every `period`th step.
+            For large values of `period` the task will be less frequently run.
+        """
+        super().__init__(period)
+        self.file_writer = tf.summary.create_file_writer(log_dir)
+
+    def __call__(self, step, **kwargs):
+        with self.file_writer.as_default():
+            super().__call__(step, **kwargs)
+        self.file_writer.flush()
+
+
+class ModelToTensorBoard(ToTensorBoard):
+    """
+    Monitoring task that creates a sensible TensorBoard for a model.
+
+    By default, it writes to the TensorBoard all model parameters that are scalars,
+    for parameter arrays (e.g. kernel.lengthscales) the values are sent to the TensorBoard
+    if the array has no more than `max_size` (default: 3) elements.
+    This behaviour can be adjusted using `max_size`.
+    """
+
+    def __init__(
+        self,
+        log_dir: str,
+        model: BayesianModel,
```

Comment:
for a BayesianModel that is not also a GPModel, this ModelToTensorBoard would by default not do anything... do you want to specify this to be a GPModel?

Reply:
To show that users can also use the monitor with their own Bayesian model. As long as the model has any `tf.Variable`s or `gpflow.Parameter`s this monitor task will do something sensible.